src/ir/auth_logic/ast.h

```diff
@@ -107,13 +128,22 @@ class CanActAs {
     return visitor.Visit(*this);
   }
 
-  // A potentially ugly print of the state in this class
-  // for debugging/testing only
-  std::string DebugPrint() const {
-    return absl::StrCat(left_principal_.DebugPrint(), " canActAs ",
-                        right_principal_.DebugPrint());
+
+  bool operator==(const CanActAs& rhs) const {
+    return left_principal_ == rhs.left_principal() &&
+           right_principal_ == rhs.right_principal();
   }
 
+  virtual bool operator==(const ASTNode& rhs) const override {
+    try {
+      const CanActAs & rhsCanActAs = dynamic_cast<const CanActAs&>(rhs);
+      return *this == rhsCanActAs;
+    } catch(const std::bad_cast& e) {
+      return false;
+    }
```

Comment:
Don't use a `try...catch` for this, as mentioned in another comment.

Reply:
fixed